src/backend/gl/src/device.rs

```diff
@@ -468,16 +469,74 @@ pub(crate) unsafe fn set_sampler_info<SetParamFloat, SetParamFloatVec, SetParamI
 impl d::Device<B> for Device {
     unsafe fn allocate_memory(
         &self,
-        _mem_type: c::MemoryTypeId,
+        mem_type: c::MemoryTypeId,
         size: u64,
     ) -> Result<n::Memory, d::AllocationError> {
-        // TODO
-        Ok(n::Memory {
-            properties: memory::Properties::CPU_VISIBLE | memory::Properties::CPU_CACHED,
-            first_bound_buffer: Cell::new(None),
-            size,
-            emulate_map_allocation: RefCell::new(std::ptr::null_mut()),
-        })
+        if (1 << mem_type.0) & IMAGE_MEM_TYPE_MASK != 0 {
+            assert_ne!((1 << mem_type.0) & DEVICE_LOCAL_MASK, 0);
+            Ok(n::Memory {
+                properties: memory::Properties::DEVICE_LOCAL,
+                buffer: None,
+                target: 0,
+                size,
+                emulate_map_allocation: Cell::new(None),
+                map_flags: 0,
+            })
+        } else {
+            assert!(self.share.private_caps.buffer_role_change);
+
+            let target = if (1 << mem_type.0) & INDEX_MEM_TYPE_MASK != 0 {
+                glow::ELEMENT_ARRAY_BUFFER
+            } else {
+                glow::ARRAY_BUFFER
+            };
+
+            let is_device_local = (1 << mem_type.0) & DEVICE_LOCAL_MASK != 0;
```

Comment:
nit: since this flag is needed in both branches, let's move this assignment up to the beginning of the function?

Reply:
Done, and I went ahead and gave all of the properties clearer names at the top of the function because I felt that was easier to follow.